2012-02-02  Aurelien Jarno  <aurel32@debian.org>

	* resolv/res_query.c(__libc_res_nsearch): succeed if the first
	query fails, but the second query succeeds.

diff --git a/resolv/res_query.c b/resolv/res_query.c
index 947c651..c88268f 100644
--- a/resolv/res_query.c
+++ b/resolv/res_query.c
@@ -378,7 +378,7 @@ __libc_res_nsearch(res_state statp,
 		ret = __libc_res_nquerydomain(statp, name, NULL, class, type,
 					      answer, anslen, answerp,
 					      answerp2, nanswerp2, resplen2);
-		if (ret > 0 || trailing_dot)
+		if (ret > 0 || (ret == 0 && *resplen2 > 0) || trailing_dot)
 			return (ret);
 		saved_herrno = h_errno;
 		tried_as_is++;
@@ -418,7 +418,7 @@ __libc_res_nsearch(res_state statp,
 						      answer, anslen, answerp,
 						      answerp2, nanswerp2,
 						      resplen2);
-			if (ret > 0)
+			if ((ret > 0) || (ret == 0 && *resplen2 > 0))
 				return (ret);
 
 			if (answerp && *answerp != answer) {
@@ -487,7 +487,7 @@ __libc_res_nsearch(res_state statp,
 		ret = __libc_res_nquerydomain(statp, name, NULL, class, type,
 					      answer, anslen, answerp,
 					      answerp2, nanswerp2, resplen2);
-		if (ret > 0)
+		if ((ret > 0) || (ret == 0 && *resplen2 > 0))
 			return (ret);
 	}
 
